src/documentation/programming-guide.md

```diff
@@ -900,8 +901,190 @@ odds = results.odd
+
+> A guide that covers how to implement your own Beam IO transforms is in progress ([BEAM-1025](https://issues.apache.org/jira/browse/BEAM-1025)).
+
+### Reading Input Data
+
+Read transforms read data from an external source and return a `PCollection` representation of the data for use by your pipeline. You can use a read transform at any point while constructing your pipeline to create a new `PCollection`, though it will be most common at the start of your program. Here are examples of two common ways to read data.
+
+#### Reading from a `Source`:
+
+```java
+// This example uses XmlSource.
+XmlSource<String> source =   XmlSource.<String>from(file.toPath().toString())
+     .withRootElement("root")
+     .withRecordElement("record")
+     .withRecordClass(Record.class);
+PCollection<String> output = pipeline.apply(Read.from(source));
+```
+
+```python
+pipeline | beam.io.Read(beam.io.TextFileSource("protocol://path/to/some/inputData.txt"))
+```
+
+Note that many sources use the builder pattern for setting options. For additional examples, see the language specific documentation (such as Javadoc) for each of the sources.
+
+#### Using a read transform:
+
+```java
+// This example uses JmsIO.
+PCollection<JmsRecord> output =
+    pipeline.apply(JmsIO.read()
+        .withConnectionFactory(myConnectionFactory)
+        .withQueue("my-queue")
+```
+
+```python
+pipeline | beam.io.textio.ReadFromText('my_file_name')
+
+```
+
+### Writing Output Data
+
+Write transforms write the data in a `PCollection` to an external data source. You will most often use write transforms at the end of your program to output your pipeline's final results. However, you can use a write transform to output a `PCollection`'s data at any point in your pipeline. Here are examples of two common ways to write data.
+
+#### Writing to a `Sink`:
+
+```java
+// This example uses XmlSink.
+pipeline.apply(Write.to(
+          XmlSink.ofRecordClass(Type.class)
+              .withRootElementName(root_element)
+              .toFilenamePrefix(output_filename)));
+```
+
+```python
+output | beam.io.Write(beam.io.TextFileSink('my_file_name'))
+```
+
+#### Using a write transform:
+
+```java
+// This example uses JmsIO.
+pipeline.apply(...) // returns PCollection<String>
+        .apply(JmsIO.write()
+            .withConnectionFactory(myConnectionFactory)
+            .withQueue("my-queue")
+```
+
+```python
+output | beam.io.textio.WriteToText('my_file_name')
+```
+
+### File-based input and output data
+
+#### Reading From Multiple Locations:
+
+Many read transforms support reading from multiple input files matching a glob operator you provide. The following TextIO example uses a glob operator (\*) to read all matching input files that have prefix "input-" and the suffix ".csv" in the given location:
+
+```java
+p.apply(“ReadFromText”,
+    TextIO.Read.from("protocol://my_bucket/path/to/input-*.csv");
+```
+
+```python
+lines = p | beam.io.Read(
```

Comment:
lines = p | beam.io.ReadFromText('protocol://my_bucket/path/to/input-*.csv')

if you want to specify a name for the transform use following.

lines = p | 'ReadFromText' >> beam.io.ReadFromText('protocol://my_bucket/path/to/input-*.csv')